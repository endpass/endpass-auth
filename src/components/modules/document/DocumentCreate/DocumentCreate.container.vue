<template>
  <document-create
    :is-closable="isClosable"
    @cancel="onClose"
  >
    <component
      :is="currentComponent"
      :document-type="documentType"
      :types="types"
      :type-to-status="typeToStatus"
      @upload="onUpload"
      @complete="handleComplete"
      @cancel="onCancel"
    />
  </document-create>
</template>

<script>
import DocumentCreate from './DocumentCreate.view';

import RequestedInfo from './modules/RequestedInfo';
import Upload from './modules/Upload';
import ExtraLoading from './modules/ExtraLoading';
import NoTypes from './modules/NoTypes';

export default {
  name: 'DocumentCreateContainer',

  props: {
    documentType: {
      type: String,
      default: '',
    },

    typeToStatus: {
      type: Object,
      required: true,
    },

    types: {
      type: Array,
      required: true,
    },

    isExtraLoading: {
      type: Boolean,
      default: false,
    },
  },

  computed: {
    isClosable() {
      // TODO: add check from store
      return !this.documentType;
    },

    currentComponent() {
      if (this.types.length === 0) {
        return NoTypes;
      }

      if (this.isExtraLoading) {
        return ExtraLoading;
      }

      if (!this.documentType) {
        return RequestedInfo;
      }

      return Upload;
    },
  },

  methods: {
    async handleComplete(payload) {
      this.$emit('complete', payload);
    },

    onClose() {
      this.$emit('close');
    },

    onCancel() {
      this.$emit('cancel');
    },

    onUpload(documentId) {
      this.$emit('create', documentId);
    },
  },

  components: {
    DocumentCreate,
  },
};
</script>
