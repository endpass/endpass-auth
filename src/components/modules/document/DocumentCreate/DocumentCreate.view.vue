<template>
  <screen @close="onCancel">
    <v-frame
      :is-closable="isClosable"
      :is-returnable="isReturnable"
      title=""
      @close="onCancel"
      @return="onReturn"
    >
      <slot />
    </v-frame>
  </screen>
</template>

<script>
import Screen from '@/components/common/Screen';
import VFrame from '@/components/common/VFrame';

export default {
  name: 'DocumentCreateView',

  props: {
    isReturnable: {
      type: Boolean,
      default: false,
    },

    isClosable: {
      type: Boolean,
      default: false,
    },
  },

  methods: {
    onReturn() {
      this.$emit('return');
    },

    onCancel() {
      this.$emit('cancel');
    },
  },

  components: {
    VFrame,
    Screen,
  },
};
</script>
