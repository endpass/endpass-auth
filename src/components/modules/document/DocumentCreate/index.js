export { default } from './DocumentCreate.interface';
