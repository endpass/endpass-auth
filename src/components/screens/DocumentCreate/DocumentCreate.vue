<template>
  <screen @close="onClose">
    <document-upload @close="onClose" />
  </screen>
</template>

<script>
import DocumentUpload from './DocumentUpload';
import Screen from '@/components/common/Screen';
import createDocumentController from './DocumentCreateController';

export default {
  name: 'DocumentCreate',

  methods: {
    onClose(isUploaded) {
      this.$options.documentCreateController.finishCreate(isUploaded);
    },
  },

  documentCreateController: createDocumentController(),

  components: {
    Screen,
    DocumentUpload,
  },
};
</script>
