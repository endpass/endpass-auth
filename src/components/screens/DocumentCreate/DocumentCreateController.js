// @ts-check
import { VuexModule, Action, Module } from 'vuex-class-modules';
// @ts-ignore
import ConnectError from '@endpass/class/ConnectError';
import createController from '@/controllers/createController';
import { documentChannel } from '@/class/singleton/channels';

import Answer from '@/class/Answer';

const { ERRORS } = ConnectError;

@Module({ generateMutationSetters: true })
class DocumentCreateController extends VuexModule {
  /**
   *
   * @param {boolean} isUploaded
   */
  @Action
  finishCreate(isUploaded) {
    const result = isUploaded
      ? Answer.createOk()
      : Answer.createFail(ERRORS.CREATE_DOCUMENT);

    documentChannel.put(result);
  }
}

export default () => createController(DocumentCreateController);
