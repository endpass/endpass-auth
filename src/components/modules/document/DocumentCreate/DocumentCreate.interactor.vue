<template>
  <document-create
    :is-returnable="isReturnable"
    :is-closable="isClosable"
    @cancel="onCancel"
    @return="onReturn"
  />
</template>

<script>
import DocumentCreate from './DocumentCreate.view';
import { coreStore, documentStore } from '@/store';

export default {
  name: 'DocumentCreate',

  coreStore,
  documentStore,

  props: {
    isReturnable: {
      type: Boolean,
      default: false,
    },
  },

  computed: {
    isClosable() {
      return this.$options.coreStore.isDialog;
    },
  },

  methods: {
    onCancel() {
      this.$emit('cancel');
    },

    onReturn() {
      this.$emit('return');
    },

    onCreate(documentId) {
      this.$emit('create', documentId);
    },

    onRequired() {
      this.$emit('required');
    },
  },

  mounted() {
    // const { clientId } = this.$options.documentStore.uploadOptions;
    // if (!clientId) {
    //   return;
    // }
    //
    // const docTypes = this.$options.documentStore.getRequiredDocumentsTypes(clientId);
  },

  components: {
    DocumentCreate,
  },
};
</script>
