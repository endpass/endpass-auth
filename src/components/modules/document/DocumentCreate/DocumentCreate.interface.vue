<template>
  <document-create
    :types="types"
    :type-to-status="typeToStatus"
    :is-extra-loading="isExtraLoading"
    :document-type="documentType"
    @complete="onComplete"
    @create="onCreate"
    @cancel="onCancel"
    @close="onClose"
  />
</template>

<script>
import DocumentCreate from './DocumentCreate.container';

export default {
  name: 'DocumentCreateInterface',

  props: {
    documentType: {
      type: String,
      default: '',
    },

    typeToStatus: {
      type: Object,
      default: () => ({}),
    },

    types: {
      type: Array,
      required: true,
    },

    isExtraLoading: {
      type: Boolean,
      default: false,
    },
  },

  methods: {
    onComplete(payload) {
      this.updateProps(payload);
    },

    updateProps(payload) {
      if (!payload) return;

      Object.keys(payload).forEach(propName => {
        this.$emit(`update:${propName}`, payload[propName]);
      });
    },

    onClose() {
      this.$emit('close');
    },

    onCancel() {
      this.$emit('cancel');
    },

    onCreate(documentId) {
      this.$emit('create', documentId);
    },
  },

  components: {
    DocumentCreate,
  },
};
</script>
