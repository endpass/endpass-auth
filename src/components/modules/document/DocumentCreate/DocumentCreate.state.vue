<template>
  <document-create
    :statuses.sync="statuses"
    :is-returnable.sync="isReturnable"
    :is-closable.sync="isClosable"
    :document-type.sync="documentType"
    @request="onRequest"
    @create="onCreate"
    @cancel="onCancel"
  />
</template>

<script>
import DocumentCreate from './DocumentCreate.container';

export default {
  name: 'DocumentCreate',

  data() {
    return {
      statuses: [],
      isReturnable: false,
      isClosable: false,
      documentType: '',
      isProcessing: false,
    };
  },

  methods: {
    onCancel() {
      this.$emit('cancel');
    },

    onCreate(documentId) {
      this.$emit('create', documentId);
    },

    onRequest() {
      this.$emit('request');
    },
  },

  components: {
    DocumentCreate,
  },
};
</script>
