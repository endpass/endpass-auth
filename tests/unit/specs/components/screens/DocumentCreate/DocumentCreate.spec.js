import VeeValidate from 'vee-validate';
import UIComponents from '@endpass/ui';

import { shallowMount, createLocalVue } from '@vue/test-utils';
import setupI18n from '@/locales/i18nSetup';
import DocumentCreate from '@/components/screens/DocumentCreate';

const localVue = createLocalVue();
const i18n = setupI18n(localVue);
localVue.use(VeeValidate);
localVue.use(UIComponents);

describe('DocumentCreate', () => {
  let wrapper;

  beforeEach(() => {
    jest.clearAllMocks();

    wrapper = shallowMount(DocumentCreate, {
      localVue,
      i18n,
      sync: false,
    });
  });

  it('should match snapshot', () => {
    expect(wrapper.html()).toMatchSnapshot();
  });
});
