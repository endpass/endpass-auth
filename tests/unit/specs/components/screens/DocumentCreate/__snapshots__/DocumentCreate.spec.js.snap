// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DocumentCreate should match snapshot 1`] = `
<screen-stub>
  <document-upload-stub></document-upload-stub>
</screen-stub>
`;
